File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
import sys
input = lambda: sys.stdin.readline().rstrip()
sys.setrecursionlimit(10**6)

def quad(y, x, size):
    if size == 1:
        return str(img[y][x])
    
    check = img[y][x]
    ret = ""

    for i in range(y, y+size):
        for j in range(x, x+size):
            if check != img[i][j]:
                ret += '('
                ret += quad(y, x, size//2)
                ret += quad(y, x+size//2, size//2)
                ret += quad(y+size//2, x, size//2)
                ret += quad(y+size//2, x+size//2 , size//2)
                ret += ')'
                return ret
    
    return str(img[y][x])

if __name__ == "__main__":
    N = int(input())
    img = [list(map(int, input())) for _ in range(N)]
    
    print(quad(0, 0, N))

File: 백준/Silver/1992. 쿼드트리/쿼드트리.cc
#include <bits/stdc++.h>
using namespace std;

const int MAX = 68;
int N; 
char a[MAX][MAX];

string quad(int y, int x, int size) {
    if (size == 1) return string(1, a[y][x]);

    char b = a[y][x];
    string ret = "";

    for (int i=y; i<y+size; i++) {
        for (int j=x; j<x+size; j++) {
            if (b != a[i][j]) {
                ret += "(";
                ret += quad(y, x, size/2);
                ret += quad(y, x+size/2, size/2);
                ret += quad(y+size/2, x, size/2);
                ret += quad(y+size/2, x+size/2, size/2);
                ret += ")";
                return ret;
            }
        }
    }
    return string(1, a[y][x]);
}

int main() {
    scanf("%d", &N);
    for (int i=0; i<N; i++) {
        scanf("%s", a[i]);
    }
    cout << quad(0, 0, N) << '\n';

    return 0;
}